### PR TITLE
fix: Add missing locals in iam-assumable-role module

### LINKED ISSUE #289: Missing partition and account_id locals in multiple sub-modules in 5.5.2
## Description

There are several local values missing in sub-modules as a consequence of merging https://github.com/terraform-aws-modules/terraform-aws-iam/pull/283.

I.e. running `terraform init && terraform plan` on 5.5.2 produces following error:
```
data.aws_iam_policy_document.custom_trust_policy: Reading...
module.iam_assumable_role_custom.data.aws_iam_policy_document.assume_role[0]: Reading...
module.iam_assumable_role_sts.data.aws_iam_policy_document.assume_role_with_mfa[0]: Reading...
data.aws_iam_policy_document.custom_trust_policy: Read complete after 1s [id=408100223]
module.iam_assumable_role_custom.data.aws_iam_policy_document.assume_role[0]: Read complete after 0s [id=2535048448]
module.iam_assumable_role_sts.data.aws_iam_policy_document.assume_role_with_mfa[0]: Read complete after 0s [id=1700556763]
module.iam_assumable_role_custom_trust_policy.data.aws_iam_policy_document.assume_role[0]: Reading...
module.iam_assumable_role_custom_trust_policy.data.aws_iam_policy_document.assume_role[0]: Read complete after 0s [id=2783130461]
╷
│ Error: Reference to undeclared local value
│
│   on ../../modules/iam-assumable-role/main.tf line 76, in data "aws_iam_policy_document" "assume_role_with_mfa":
│   76:         values   = ["arn:${local.partition}:iam::${local.account_id}:role${var.role_path}${local.role_name_condition}"]
│
│ A local value with the name "partition" has not been declared.
╵
╷
│ Error: Reference to undeclared local value
│
│   on ../../modules/iam-assumable-role/main.tf line 76, in data "aws_iam_policy_document" "assume_role_with_mfa":
│   76:         values   = ["arn:${local.partition}:iam::${local.account_id}:role${var.role_path}${local.role_name_condition}"]
│
│ A local value with the name "account_id" has not been declared.
╵
```

- [X] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version 5.5.2

- Terraform version: ANY
- Provider version(s): ANY

## Reproduction Code [Required]

* Go to `examples/iam-assumable-role`
* Run `terraform init && terraform plan`

See bove

## Expected behavior

Terrform should produce a clean plan

## Actual behavior

It fails with error A local value with the name "partition" and "account_id" has not been declared.

## Additional context

See https://github.com/terraform-aws-modules/terraform-aws-iam/pull/283

### PR DESCRIPTION
**Pull request by enver**
## Description
Fixes #289 

## Motivation and Context
Fixes #289 

## Breaking Changes
No

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [X] I have tested and validated these changes using one or more of the provided `examples/*` projects
- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

**Comment by sirantd** [comment]
Please take a look at line No. 81 of the main.tf, I believe we need to keep consistency and either use `local.` in both places or not use them in both places.

**Comment by enver** [comment]
@sirantd I'm don't understand your comment. Could you please add more details?

**Comment by sirantd** [comment]
@enver It has been fixed in your second commit, all good now

image

**Comment by LucaIcaro** [comment]
Can this be approved? We have some workflows that are blocked because of this issue.

**Comment by sirantd** [comment]
I made [PR](https://github.com/enver/terraform-aws-iam/pull/1) to the @enver branch. Once merged - it looks ok to me.

**Comment by enver** [comment]
@sirantd Sorry, I just couldn't find time to finish this. Thank you for jumping in...merged.

**Comment by sirantd** [comment]
Hi @bryantbiggs, does it makes sense to do fixes to pass linter warnings for terraform 1.3.3 as part of this PR? It's already quite big. 

Could you review it so it can be merged, please?

**Comment by tegail** [comment]
can we get someone to review this? it is also blocking workflows for us as well

**Comment by saurabh-paystack** [comment]
When can we expect this to be merged, please?

File: examples/iam-assumable-roles-with-saml/main.tf [hunk 1]
```diff
@@ -58,3 +58,18 @@ module "iam_assumable_roles_with_saml_custom" {
 
   provider_id = aws_iam_saml_provider.idp_saml.id
 }
+
+################################################
+# IAM assumable roles with SAML with self assume
+################################################
+module "iam_assumable_roles_with_saml_with_self_assume" {
```
Comment (antonbabenko) [inline examples/iam-assumable-roles-with-saml/main.tf:65]: ALL examples should be executable right away, as described in the README - `terraform init -upgrade && terraform apply -auto-approve && terraform destroy -auto-approve`.

Currently, some of the examples are failing with `EntityAlreadyExists` errors. Please fix.

File: modules/iam-assumable-role-with-oidc/main.tf [hunk 2]
```diff
@@ -32,7 +33,7 @@ data "aws_iam_policy_document" "assume_role_with_oidc" {
       condition {
         test     = "ArnLike"
         variable = "aws:PrincipalArn"
-        values   = ["arn:${local.partition}:iam::${local.account_id}:role${var.role_path}${local.role_name_condition}"]
+        values   = ["arn:${local.partition}:iam::${data.aws_caller_identity.current.account_id}:role${var.role_path}${local.role_name_condition}"]
```
Comment (antonbabenko) [inline modules/iam-assumable-role-with-oidc/main.tf:36]: Revert this change.

File: modules/iam-assumable-role-with-oidc/main.tf [hunk 3]
```diff
@@ -47,7 +48,7 @@ data "aws_iam_policy_document" "assume_role_with_oidc" {
       principals {
         type = "Federated"
 
-        identifiers = ["arn:${data.aws_partition.current.partition}:iam::${local.aws_account_id}:oidc-provider/${statement.value}"]
+        identifiers = ["arn:${local.partition}:iam::${local.aws_account_id}:oidc-provider/${statement.value}"]
```
Comment (antonbabenko) [inline modules/iam-assumable-role-with-oidc/main.tf:51]: Also this one.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Minor fixes and, most importantly, all examples should be working.

File: modules/iam-assumable-role-with-oidc/main.tf (hunk 2 above)
Comment (sirantd) [reply modules/iam-assumable-role-with-oidc/main.tf:36]: Could you please elaborate on why it's better to use data instead of local?

Comment (antonbabenko) [reply modules/iam-assumable-role-with-oidc/main.tf:36]: Usually, do not touch it if it works is a very important rule. Such microchanges pollute the PR and make it hard to review. This PR is doing more than fixing an original issue.

Introducing `locals` makes perfect sense when there is some logic in it (e.g. `local.aws_account_id`).

Comment (sirantd) [reply modules/iam-assumable-role-with-oidc/main.tf:36]: Got it, thank you for the explanation.

Comment (enver) [reply modules/iam-assumable-role-with-oidc/main.tf:36]: @antonbabenko  This one can't be reverted because the module will stop working due to missing `account_id` local. If you check locals above you can see that `aws_account_id` is declared, however, `account_id` is not. When I looked at [this](https://github.com/terraform-aws-modules/terraform-aws-iam/pull/283/files#diff-ab579ba376417161fb0950cae991b5ffb34ee5c20c780f0cc154ad5fbf263880L28) change which introduced the problem it seemed obvious to just revert it to state it was before.

File: examples/iam-assumable-roles-with-saml/main.tf (hunk 1 above)
Comment (enver) [reply examples/iam-assumable-roles-with-saml/main.tf:67]: @antonbabenko Fixed most of the examples that were failing with error above. NOTE: `iam-eks-role` can't be applied for some time now so I consider it out of the scope to fix it in this PR.

File: modules/iam-assumable-role-with-oidc/main.tf (hunk 3 above)
Comment (enver) [reply modules/iam-assumable-role-with-oidc/main.tf]: Reverted

**Review by antonbabenko — APPROVED** [review]
Looks good. As said, examples/iam-eks-role is failing, but it is not related to this PR.

**Comment by antonbabenko** [comment]
This PR is included in version 5.5.5 :tada:
